README.rst
================
Welcome to eyeD3
================

Python audio data toolkit (ID3 and MP3)

Status
------
.. image:: https://img.shields.io/pypi/v/eyeD3.svg
   :target: https://pypi.python.org/pypi/eyeD3/
   :alt: Latest Version
.. image:: https://img.shields.io/pypi/status/eyeD3.svg
   :target: https://pypi.python.org/pypi/eyeD3/
   :alt: Project Status
.. image:: https://travis-ci.org/nicfit/eyeD3.svg?branch=master
   :target: https://travis-ci.org/nicfit/eyeD3
   :alt: Build Status
.. image:: https://img.shields.io/pypi/l/eyeD3.svg
   :target: https://pypi.python.org/pypi/eyeD3/
   :alt: License
.. image:: https://img.shields.io/pypi/pyversions/eyeD3.svg
   :target: https://pypi.python.org/pypi/eyeD3/
   :alt: Supported Python versions
.. image:: https://coveralls.io/repos/nicfit/eyeD3/badge.svg
   :target: https://coveralls.io/r/nicfit/eyeD3
   :alt: Coverage Status


About
-----
eyeD3_ is a Python tool for working with audio files, specifically mp3 files
containing ID3_ metadata (i.e. song info).

It provides a command-line tool (``eyeD3``) and a Python library
(``import eyed3``) that can be used to write your own applications or
plugins that are callable from the command-line tool.

For example, to set some song information in an mp3 file called
``song.mp3``::

  $ eyeD3 -a Nobunny -A "Love Visions" -t "I Am a Girlfriend" -n 4 song.mp3

With this command we've set the artist (``-a/--artist``), album
(``-A/--album``), title (``-t/--title``), and track number
(``-n/--track-num``) properties in the ID3 tag of the file. This is the
standard interface that eyeD3 has always had in the past, therefore it
is also the default plugin when no other is specified.

The results of this command can be seen by running the ``eyeD3`` with no
options.

::

  $ eyeD3 song.mp3
  song.mp3	[ 3.06 MB ]
  -------------------------------------------------------------------------
  ID3 v2.4:
  title: I Am a Girlfriend
  artist: Nobunny
  album: Love Visions
  album artist: Various Artists
  track: 4		
  -------------------------------------------------------------------------
  
The same can be accomplished using Python.

::

  import eyed3

  audiofile = eyed3.load("song.mp3")
  audiofile.tag.artist = u"Nobunny"
  audiofile.tag.album = u"Love Visions"
  audiofile.tag.album_artist = u"Various Artists"
  audiofile.tag.title = u"I Am a Girlfriend"
  audiofile.tag.track_num = 4

  audiofile.tag.save()

eyeD3_ is written and maintained by `Travis Shirk`_ and is licensed under
version 2 of the GPL_.

Features
--------

* Python package for writing application and/or plugins.
* Command-line tool driver script that supports plugins.
  viewer/editor interface.
* Easy editing/viewing of audio metadata from the command-line, using the
  'classic' plugin.
* Support for ID3 versions 1.x, 2.2 (read-only), 2.3, and 2.4.
* Support for the MP3 audio format exposing details such as play time, bit
  rate, sampling frequency, etc.
* Abstract design allowing future support for different audio formats and
  metadata containers.


Get Started
-----------

Python 2.6 or 2.7 (recommended) is required.

For `installation instructions`_ or more complete `documentation`_ see
http://eyeD3.nicfit.net/

Please post feedback and/or defects on the `issue tracker`_, or `mailing list`_.

.. _eyeD3: http://eyeD3.nicfit.net/
.. _Travis Shirk: travis@pobox.com
.. _issue tracker: https://bitbucket.org/nicfit/eyed3/issues?status=new&status=open
.. _mailing list: https://groups.google.com/forum/?fromgroups#!forum/eyed3-users
.. _installation instructions: http://eyeD3.nicfit.net/index.html#installation
.. _documentation: http://eyeD3.nicfit.net/index.html#documentation
.. _GPL: http://www.gnu.org/licenses/gpl-2.0.html
.. _ID3: http://id3.org/



ChangeLog
commit 0059b6e0b1cea3fd90c98fb1dbc3fc69e9c3d923
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 16:05:46 2017 -0600

    Release test

commit f6ac88a43f05b1e62fdd01664ddffbb2a4b3da50
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 16:05:00 2017 -0600

    Release testing

commit 24692a5b00e2e2f3bf5202440999177a5c7fb2ad
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 16:02:59 2017 -0600

    Release tests.  !wip

commit 2006d328e2fc3584e2650d196391f03750c62ffd
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 15:57:39 2017 -0600

    Travis-CI 0.7 build testing. !wip

commit 6b50c9df1f8f352c45639e475e2d41bfa6b7d33c
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 15:44:52 2017 -0600

    0.7.11 release prep.  !wip

commit 8bec4ce8d0d9004ca57deed70884f9c3894bfd00
Author: Travis Shirk <travis@pobox.com>
Date:   Sun Mar 12 13:02:30 2017 -0600

    fix: Replace unicode during v1.x conversion. Fixes BB #125

commit 6b509ef7eb3e1f15be9c939f43a73a035082afdc
Author: nicfit <travis@pobox.com>
Date:   Wed Jan 4 19:53:06 2017 -0700

    version/date bump

commit 2528b7b31528d6aee30761ec3a7f9c06cda3f84a
Author: nicfit <travis@pobox.com>
Date:   Tue Dec 27 19:12:36 2016 -0700

    More to stdout and less to stderr. Fixes #113, for the most part, reopen if not.

commit df40bbd36fcb55d01a2b2a9ad08b1fb0a6d713ce
Author: nicfit <travis@pobox.com>
Date:   Tue Dec 27 16:44:16 2016 -0700

    Updated

commit 77cfafca2e46d595be480b9bef7b2b46f8d11e64
Author: nicfit <travis@pobox.com>
Date:   Tue Dec 27 16:05:58 2016 -0700

    Allow --no-tagging-time-frame. Fixes #110

commit e15495bed9f8b0caaf304a006a2050a2d7c7f63d
Author: nicfit <travis@pobox.com>
Date:   Tue Dec 27 13:57:41 2016 -0700

    Date fixes, and added back deprecated --year opt. Fixes #108 and #116

commit 07fa09ece06185938793167d7f1052ed98e820b0
Author: nicfit <travis@pobox.com>
Date:   Tue Dec 27 12:55:46 2016 -0700

    Fix #117: --user-text-frame= parse error for multiple colons

commit 20c0b389a157c91c26d6003b66d420158667e6dc
Merge: 75dcd9b 642ac90
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 14:50:47 2016 -0700

    merged default

commit 642ac90947d11fc7555c24d448e75001eab64ff1
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 14:30:48 2016 -0700

    Added tag v0.7.10 for changeset 3e030e1b4f32

commit 302ae13b03b7ba5a4bdfb0fb414d3640f42a817b
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 14:30:09 2016 -0700

    prep for release

commit eda8e5dac794e020542b98004984d627b8251de4
Merge: 75dcd9b 4942262
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 14:28:35 2016 -0700

    merge 0.7.x

commit 75dcd9bbb25b65ffac1f35412ef6133410f5695b
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 14:28:10 2016 -0700

    Release prep

commit 494226214b0c9de92b78d916adfa1a9ca4ddb9e9
Merge: 6bcde2c 372e525
Author: nicfit <travis@pobox.com>
Date:   Sat Dec 10 11:34:31 2016 -0700

    Merge 0.7.x branch

commit 6bcde2c0ae12e59c9164ab4e46243c319563e88c
Author: nicfit <travis@pobox.com>
Date:   Wed Nov 30 22:27:34 2016 -0700

    Backport of #105 fix

commit 67a93ddf0c5c33c56b13166b5c9b8a660d3290c4
Author: nicfit <travis@pobox.com>
Date:   Wed Nov 30 22:24:28 2016 -0700

    Fixed frame counts in debug output

commit bfd4c8a81d87dd2898388a3ef9d4180880411fdd
Author: nicfit <travis@pobox.com>
Date:   Wed Nov 30 19:00:06 2016 -0700

    Fixed Tag.frameiter()

commit 592a465b64dd1ba13d9f6001c28dc590d9a0c58e
Author: nicfit <travis@pobox.com>
Date:   Wed Feb 17 18:13:17 2016 -0700

    Import log before using it, fixes #97

commit c6f36f961f925bec6b017dc65b78409df52c3c89
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 16:22:22 2015 -0700

    Branch for 0.7.x maintenance

commit 372e525df9ca9bad11fb62b050b4d9ab449ade83
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 15:57:28 2015 -0700

    Added tag v0.7.9 for changeset c22b00531324

commit 412957dca569daf974eb0167e9f7127c168cb2ac
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 15:57:03 2015 -0700

    prep for release

commit 2fc07d0939f59c4002fe492b4a633c65efee5de4
Merge: 83b4c0a bea65f3
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 15:42:50 2015 -0700

    Merged devel

commit 83b4c0a9c2ac9a1ec6f5742f169091b61d6b793f
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 15:41:03 2015 -0700

    release prep

commit c75e53449bed46ec15ce41319802a33d43a9f512
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 15:23:38 2015 -0700

    changelog updates, and whitespace

commit 89e0edccbf9c167e87037f3f15280f334af73457
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 14:55:03 2015 -0700

    prep

commit 27175577c89942cdf3dfcad8aaf25a0e117a9952
Author: nicfit <travis@pobox.com>
Date:   Thu Nov 26 14:54:27 2015 -0700

    Handle/convert unicode Image mime-types

commit c2c8a02462e788913051e25a6a37890759eda57a
Author: nicfit <travis@pobox.com>
Date:   Sat Nov 21 09:45:24 2015 -0700

    Merged 'default' branch

commit bea65f3a8601a5ab9e6345fc884a739d7fe8cfbb
Author: Hans Petter Jansson <hpj@copyleft.no>
Date:   Wed Nov 11 17:13:18 2015 +0100

    sort frames by id, which keeps the order consistent _across invocations_

commit c8eedc07be35c82a33094a42e8f0c85a4b3be1ae
Author: Hans Petter Jansson <hpj@copyleft.no>
Date:   Wed Nov 11 10:42:53 2015 +0100

    some code has problems with image descriptions encoded <> latin1
    namely mp3diags: work around the problem by forcing latin1 encoding for
    empty descriptions, which is by far the most common case anyway

commit 1b434b159742f19d0b1b3fa8f14cc5c9b62e1e7a
Author: Hans Petter Jansson <hpj@copyleft.no>
Date:   Tue Oct 20 13:44:57 2015 +0200

    stating to be unable to find a valid mp3 frame without a hint,
    _where_ this happened is rather unfortunate.
    I noticed this from using eyed3.load() calls.

commit 599eddd2b347fa32bfb524115770f3377636fd5d
Author: Hans Petter Jansson <hpj@copyleft.no>
Date:   Tue Oct 20 13:44:06 2015 +0200


pavement.py
# -*- coding: utf-8 -*-
################################################################################
#  Copyright (C) 2012-2015  Travis Shirk <travis@pobox.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>
################################################################################
from __future__ import print_function
import os
import re
from paver.easy import *
from paver.path import path
import paver.setuputils
import paver.doctools
paver.setuputils.install_distutils_tasks()
import setuptools
import setuptools.command
try:
    from sphinxcontrib import paverutils
except:
    paverutils = None

PROJECT = u"eyeD3"
VERSION = "0.7.11"

LICENSE = open("COPYING", "r").read().strip('\n')
DESCRIPTION = "Python audio data toolkit (ID3 and MP3)"
LONG_DESCRIPTION = """
eyeD3 is a Python module and command line program for processing ID3 tags.
Information about mp3 files (i.e bit rate, sample frequency,
play time, etc.) is also provided. The formats supported are ID3
v1.0/v1.1 and v2.3/v2.4.
"""
URL = "http://eyeD3.nicfit.net/"
AUTHOR = "Travis Shirk"
AUTHOR_EMAIL = "travis@pobox.com"
SRC_DIST_TGZ = "%s-%s.tar.gz" % (PROJECT, VERSION)
SRC_DIST_ZIP = "%s.zip" % SRC_DIST_TGZ.replace(".tar.gz", "")
DOC_DIST = "%s_docs-%s.tar.gz" % (PROJECT, VERSION)
DOC_BUILD_D = "docs/_build"

PACKAGE_DATA = paver.setuputils.find_package_data("src/eyed3",
                                                  package="eyed3",
                                                  only_in_packages=True,
                                                  )
DEPS = []

options(
    minilib=Bunch(
        extra_files=['doctools', "shell"]
    ),
    setup=Bunch(
        name=PROJECT, version=VERSION,
        description=DESCRIPTION, long_description=LONG_DESCRIPTION,
        author=AUTHOR, maintainer=AUTHOR,
        author_email=AUTHOR_EMAIL, maintainer_email=AUTHOR_EMAIL,
        url=URL,
        download_url="%s/releases/%s" % (URL, SRC_DIST_TGZ),
        license="GPL",
        package_dir={"": "src"},
        packages=setuptools.find_packages("src",
                                          exclude=["test", "test.*"]),
        zip_safe=False,
        classifiers = [
            'Environment :: Console',
            'Intended Audience :: End Users/Desktop',
            'Intended Audience :: Developers',
            'License :: OSI Approved :: GNU General Public License v2 (GPLv2)',
            'Operating System :: POSIX',
            'Programming Language :: Python',
            'Programming Language :: Python :: 2.6',
            'Programming Language :: Python :: 2.7',
            'Programming Language :: Python :: 2 :: Only',
            'Topic :: Multimedia :: Sound/Audio :: Editors',
            'Topic :: Software Development :: Libraries :: Python Modules',
            ],
        platforms=["Any",],
        keywords=["id3", "mp3", "python"],
        scripts=["bin/eyeD3"],
        package_data=PACKAGE_DATA,
        install_requires=DEPS,
    ),

    sdist=Bunch(
        formats="gztar,zip",
        dist_dir="dist",
    ),

    sphinx=Bunch(
        docroot=os.path.split(DOC_BUILD_D)[0],
        builddir=os.path.split(DOC_BUILD_D)[1],
        builder='html',
        template_args = {},
    ),

    cog=Bunch(
        beginspec='{{{cog',
        endspec='}}}',
        endoutput='{{{end}}}',
        includedir=path(__file__).abspath().dirname(),
    ),

    test=Bunch(
        pdb=False,
        coverage=False,
    ),

    release=Bunch(
        test=False,
    ),
)


@task
@no_help
def eyed3_info():
    '''Convert src/eyed3/info.py.in to src/eyed3/info.py'''
    src = path("./src/eyed3/info.py.in")
    target = path("./src/eyed3/info.py")
    if target.exists() and not src.exists():
        return
    elif not src.exists():
        raise Exception("Missing src/eyed3/info.py.in")
    elif not target.exists() or src.ctime > target.ctime:
        src_file = src.open("r")
        target_file = target.open("w")

        src_data = re.sub("@PROJECT@", PROJECT, src_file.read())
        src_data = re.sub("@VERSION@", VERSION.split('-')[0], src_data)
        src_data = re.sub("@AUTHOR@", AUTHOR, src_data)
        src_data = re.sub("@URL@", URL, src_data)
        if '-' in VERSION:
            src_data = re.sub("@RELEASE@", VERSION.split('-')[1], src_data)
        else:
            src_data = re.sub("@RELEASE@", "final", src_data)

        target_file.write(src_data)
        target_file.close()


@task
@needs("eyed3_info",
       "generate_setup",
       "minilib",
       "setuptools.command.build")
def build():
    '''Build the code'''
    pass


@task
@needs("test_clean")
def clean():
    '''Cleans mostly everything'''
    path("build").rmtree()

    for p in path(".").glob("*.pyc"):
        p.remove()
    for d in [path("./src"), path("./examples")]:
        for f in d.walk(pattern="*.pyc"):
            f.remove()
    try:
        from paver.doctools import uncog
        uncog()
    except ImportError:
        pass

    sh("git checkout paver-minilib.zip")


@task
def docs_clean(options):
    '''Clean docs'''
    for d in ["html", "doctrees"]:
        path("%s/%s" % (DOC_BUILD_D, d)).rmtree()

    try:
        from paver.doctools import uncog
        uncog()
    except ImportError:
        pass


@task
@needs("distclean", "docs_clean", "tox_clean")
def maintainer_clean():
    path("paver-minilib.zip").remove()
    path("setup.py").remove()
    path("src/eyed3/info.py").remove()


@task
@needs("clean")
def distclean():
    '''Like 'clean' but also everything else'''
    path("tags").remove()
    path("dist").rmtree()
    path("src/eyeD3.egg-info").rmtree()
    for pat in ("*.orig", "*.rej"):
        for f in path(".").walk(pattern=pat):
            f.remove()
    path(".ropeproject").rmtree()


@task
@needs("cog")
def docs(options):
    '''Sphinx documenation'''
    if not paverutils:
        raise RuntimeError("Sphinxcontib.paverutils needed to make docs")
    sh("sphinx-apidoc --force -o ./docs/api ./src/eyed3/")
    paverutils.html(options)
    print("Docs: file://%s/%s/%s/html/index.html" %
          (os.getcwd(), options.docroot, options.builddir))


@task
@needs("distclean",
       "eyed3_info",
       "generate_setup",
       "minilib",
       "setuptools.command.sdist",
       )
def sdist(options):
    '''Make a source distribution'''
    cwd = os.getcwd()
    try:
        name = SRC_DIST_TGZ.replace(".tar.gz", "")
        os.chdir(options.sdist.dist_dir)
        # Caller of sdist can select the type of output, so existence checks...
    finally:
        os.chdir(cwd)


@task
def tox(options):
    sh("tox")


@task
def tox_clean(options):
    sh("rm -rf .tox")


@task
def changelog():
    '''Update changelog, and commit it'''
    sh("git log >| ChangeLog")


@task
@no_help
def tags():
    '''ctags for development'''
    path("tags").remove()
    sh("ctags -R --exclude='tmp/*' --exclude='build/*' ./src")


@task
@needs("build")
@cmdopts([("pdb", "",
           u"Run with all output and launch pdb for errors and failures"),
          ("coverage", "", u"Run tests with coverage analysis"),
         ])
def test(options):
    '''Runs all tests'''
    if options.test and options.test.pdb:
        debug_opts = "--pdb --pdb-failures -s"
    else:
        debug_opts = ""

    if options.test and options.test.coverage:
        coverage_opts = (
            "--cover-erase --with-coverage --cover-tests --cover-inclusive "
            "--cover-package=eyed3 --cover-branches --cover-html "
            "--cover-html-dir=build/test/coverage src/test")
    else:
        coverage_opts = ""

    sh("nosetests --verbosity=1 --detailed-errors "
       "%(debug_opts)s %(coverage_opts)s" %
       {"debug_opts": debug_opts, "coverage_opts": coverage_opts})

    if coverage_opts:
        print("Coverage Report: file://%s/build/test/coverage/index.html" %
              os.getcwd())


@task
def test_clean():
    '''Clean tests'''
    path("built/test/html").rmtree()
    path(".coverage").remove()


@task
@needs("sdist")
def test_dist():
    '''Makes a dist package, unpacks it, and tests it.'''
    cwd = os.getcwd()
    pkg_d = SRC_DIST_TGZ.replace(".tar.gz", "")

    try:
        os.chdir("./dist")
        sh("tar xzf %s" % SRC_DIST_TGZ)

        os.chdir(pkg_d)
        # Copy tests into src pkg
        sh("cp -r ../../src/test ./src")
        sh("python setup.py build")
        sh("python setup.py test")

        os.chdir("..")
        path(pkg_d).rmtree()
    finally:
        os.chdir(cwd)


@task
@needs("docs")
def docdist():
    path("./dist").exists() or os.mkdir("./dist")
    cwd = os.getcwd()
    try:
        os.chdir(DOC_BUILD_D)
        sh("tar czvf ../../dist/%s html" % DOC_DIST)
        os.chdir("%s/dist" % cwd)
    finally:
        os.chdir(cwd)

    pass


@task
@cmdopts([("test", "",
           u"Run in a mode where commits, pushes, etc. are not performed"),
         ])
def release(options):
    from paver.doctools import uncog

    testing = options.release.test
    if testing:
        print("** Release testing mode **")

    # Ensure we're on default branch
    sh("test $(git rev-parse --abbrev-ref HEAD) = '0.7.x'")

    if not prompt("Is version *%s* correct?" % VERSION):
        print("Fix VERSION")
        return

    if not prompt("Is docs/changelog.rst up to date?"):
        print("Update changlelog")
        return

    print("Checking for clean working copy")
    if not testing:
        sh("git diff --quiet && git diff --quiet --staged")

    changelog()
    if prompt("Commit ChangeLog?") and not testing:
        sh("git add -u && git commit -n -m 'prep for release'")

    test()
    tox()

    sdist()
    docdist()
    uncog()
    test_dist()

    # Undo this lame update
    sh("git checkout paver-minilib.zip")

    if prompt("Tag release 'v%s'?" % VERSION) and not testing:
        sh("git tag -a v%s -m 'Release v%s'" % (VERSION, VERSION))
        sh("git push --tags origin")

    if prompt("Push for release?") and not testing:
        sh("git push")


def prompt(prompt):
    print(prompt + ' ', end='')
    resp = raw_input()
    return True if resp in ["y", "yes"] else False


def cog_pluginHelp(name):
    from string import Template
    import argparse
    import eyed3.plugins

    substs = {}
    template = Template(
'''
*$summary*

Names
-----
$name $altnames

Description
-----------
$description

Options
-------
.. code-block:: text

$options

''')

    plugin = eyed3.plugins.load(name)
    substs["name"] = plugin.NAMES[0]
    if len(plugin.NAMES) > 1:
        substs["altnames"] = "(aliases: %s)" % ", ".join(plugin.NAMES[1:])
    else:
        substs["altnames"] = ""
    substs["summary"] = plugin.SUMMARY
    substs["description"] = plugin.DESCRIPTION if plugin.DESCRIPTION else u""

    arg_parser = argparse.ArgumentParser()
    _ = plugin(arg_parser)

    buffer = u""
    found_opts = False
    for line in arg_parser.format_help().splitlines(True):
        if not found_opts:
            if (line.lstrip().startswith('-') and
                    not line.lstrip().startswith("-h")):
                buffer += (" " * 2) + line
                found_opts = True
        else:
            if buffer == '\n':
                buffer += line
            else:
                buffer += (" " * 2) + line
    if buffer.strip():
        substs["options"] = buffer
    else:
        substs["options"] = u"  No extra options supported"

    return template.substitute(substs)
__builtins__["cog_pluginHelp"] = cog_pluginHelp


# XXX: modified from paver.doctools._runcog to add includers
def _runcog(options, uncog=False):
    """Common function for the cog and runcog tasks."""

    eyed3_info()

    import cogapp
    options.order('cog', 'sphinx', add_rest=True)
    c = cogapp.Cog()
    if uncog:
        c.options.bNoGenerate = True
    c.options.bReplace = True
    c.options.bDeleteCode = options.get("delete_code", False)
    includedir = options.get('includedir', None)
    if includedir:
        markers = options.get("include_markers")

        include = Includer(includedir, cog=c,
                           include_markers=options.get("include_markers"))
        # load cog's namespace with our convenience functions.
        c.options.defines['include'] = include
        c.options.defines['sh'] = _cogsh(c)

        cli_includer = CliExample(includedir, cog=c, include_markers=markers)
        c.options.defines["cli_example"] = cli_includer

    c.options.defines.update(options.get("defines", {}))

    c.sBeginSpec = options.get('beginspec', '[[[cog')
    c.sEndSpec = options.get('endspec', ']]]')
    c.sEndOutput = options.get('endoutput', '[[[end]]]')

    basedir = options.get('basedir', None)
    if basedir is None:
        basedir = path(options.get('docroot', "docs")) / \
                  options.get('sourcedir', "")
    basedir = path(basedir)

    pattern = options.get("pattern", "*.rst")
    if pattern:
        files = basedir.walkfiles(pattern)
    else:
        files = basedir.walkfiles()
    for f in files:
        dry("cog %s" % f, c.processOneFile, f)

from paver.doctools import Includer, _cogsh


class CliExample(Includer):
    def __call__(self, fn, section=None, lang="bash"):
        # Resetting self.cog to get a string back from Includer.__call__
        cog = self.cog
        self.cog = None
        raw = Includer.__call__(self, fn, section=section)
        self.cog = cog

        self.cog.cogmodule.out(u"\n.. code-block:: %s\n\n" % lang)
        for line in raw.splitlines(True):
            if line.strip() == "":
                self.cog.cogmodule.out(line)
            else:
                cmd = line.strip()
                cmd_line = ""
                if not cmd.startswith('#'):
                    cmd_line = "$ %s\n" % cmd
                else:
                    cmd_line = cmd + '\n'

                cmd_line = (' ' * 2) + cmd_line
                self.cog.cogmodule.out(cmd_line)

                if cmd.startswith("eyeD3 "):
                    cmd += " --no-color --no-config "
                output = sh(cmd, capture=True)
                if output:
                    self.cog.cogmodule.out("\n")
                for ol in output.splitlines(True):
                    self.cog.cogmodule.out(' ' * 2 + ol)
                if output:
                    self.cog.cogmodule.out("\n")


@task
def cog(options):
    '''Run cog on all docs'''
    _runcog(options)


TEST_DATA_FILE = "eyeD3-test-data.tgz"
TEST_DATA_D = os.path.splitext(TEST_DATA_FILE)[0]

@task
def test_data(options):
    cwd = os.getcwd()

    sh("wget --quiet 'http://nicfit.net/files/%(TEST_DATA_FILE)s'" % globals())
    sh("tar xzf ./%(TEST_DATA_FILE)s -C ./src/test" % globals())
    try:
        os.chdir("./src/test")
        sh("ln -s ./%(TEST_DATA_D)s ./data" % globals())
    finally:
        os.chdir(cwd)

@task
def test_data_clean(options):
    sh("rm ./%(TEST_DATA_FILE)s" % globals())
    if os.path.lexists("src/test/data"):
        sh("rm src/test/data")
    sh("rm -rf src/test/%(TEST_DATA_D)s" % globals(), ignore_error=True)
